File: src/gatling/java/com/assignment/product_service/performance/ProductServiceSimulation.java
package com.assignment.product_service.performance;



import static io.gatling.javaapi.core.CoreDsl.StringBody;
import static io.gatling.javaapi.core.CoreDsl.atOnceUsers;
import static io.gatling.javaapi.core.CoreDsl.rampUsers;
import static io.gatling.javaapi.core.CoreDsl.scenario;
import static io.gatling.javaapi.http.HttpDsl.http;
import static io.gatling.javaapi.http.HttpDsl.status;

import io.gatling.javaapi.core.ScenarioBuilder;
import io.gatling.javaapi.core.Simulation;
import io.gatling.javaapi.http.HttpProtocolBuilder;
import io.gatling.javaapi.http.HttpRequestActionBuilder;

public class ProductServiceSimulation extends Simulation {

    // HTTP Protocol Configuration
    HttpProtocolBuilder httpProtocol = http
        .baseUrl("http://localhost:8082")
        .acceptHeader("application/json")
        .contentTypeHeader("application/json");

    // JSON Payload for the GET request (if needed)
    // Though GET APIs typically do not have a body, you can customize headers or query parameters as needed

    // Define the GET request for retrieving items
    HttpRequestActionBuilder getItemsRequest = http("Get Items")
        .get("/api/items")
        .check(status().is(200)); // Check that the status code is 200

    // Define the scenario: Send requests to get items
    ScenarioBuilder scn = scenario("Items Retrieval Load Test")
        .exec(getItemsRequest);

    {
        setUp(
            scn.injectOpen(
                atOnceUsers(50),          // Inject 5 users at once (concurrent requests)
                rampUsers(100).during(10) // Ramp up to 10 users over 10 seconds
            ).protocols(httpProtocol)
        );
    }
}

File: src/gatling/scala/com/assignment/performance/ProductServiceSimulation.scala
package com.assignment.performance

import io.gatling.core.Predef._
import io.gatling.http.Predef._
import scala.concurrent.duration._

class ProductServiceSimulation extends Simulation {

  // HTTP Protocol Configuration
  val httpProtocol = http
    .baseUrl("http://localhost:8082") // Base URL for the requests
    .acceptHeader("application/json") // Accept header
    .contentTypeHeader("application/json") // Content-type header for POST requests

  // JSON Payload for POST request
  val createProductPayload = """{
    "name": "Test Product",
    "quantity" : 10
  }"""

  // Scenario: Create Product Load Test
  val createProductScenario = scenario("Create Product Load Test")
    .exec(
      http("Create Product")
        .post("/api/items")
        .body(StringBody(createProductPayload)) // Send the JSON payload
        .asJson
        .check(status.is(200)) // Check that the status code is 200 (Created)
    )

  // Load Simulation Configuration
  setUp(
    createProductScenario.inject(
      // Inject users into the simulation
      atOnceUsers(10), // Inject 10 users immediately
      rampUsers(10).during(10.seconds), // Ramp up to 10 users over 10 seconds
      constantUsersPerSec(2).during(30.seconds) // Constant 2 users per second for 30 seconds
    ).protocols(httpProtocol)
  )

  // Optional: Throttle configuration
  .throttle(
    // Limit to 10 requests per second
    reachRps(10).in(10.seconds),
    // Hold 10 requests per second for 30 seconds
    holdFor(30.seconds),
    // Drop to 5 requests per second
    jumpToRps(5),
    // Hold for 10 seconds
    holdFor(10.seconds)
  )
}
